src/xref2/env.ml

```diff
@@ -80,97 +80,97 @@ type kind =
-    let names = StringMap.add name ({ kind; elem } :: tl) names in
-    (names, ids)
-
-  let remove identifier (names, ids) =
-    let id = (identifier :> Identifier.t) in
-    let elem =
-      try IdMap.find id ids
-      with Not_found ->
-        Format.eprintf "Failed to find %a\n%!" Component.Fmt.model_identifier id;
-        assert false
-    in
-    let names' =
-      let name = Identifier.name id in
-      let l = StringMap.find name names in
-      match
-        List.filter (fun e -> not Component.Element.(equal elem e.elem)) l
-      with
-      | [] -> StringMap.remove name names
-      | xs -> StringMap.add name xs (StringMap.remove name names)
+    StringMap.add name ({ kind; elem } :: tl) t
+
+  let remove id t =
+    let id = (id :> Identifier.t) in
+    let name = Identifier.name id in
+    let l = StringMap.find name t in
+    match
+      List.filter
+        (fun e ->
+          not (Identifier.equal id (Component.Element.identifier e.elem)))
+        l
+    with
+    | [] -> StringMap.remove name t
+    | xs -> StringMap.add name xs (StringMap.remove name t)
+
+  let find_by_name f name t =
+    let filter e acc = match f e.elem with Some r -> r :: acc | None -> acc in
+    try List.fold_right filter (StringMap.find name t) [] with Not_found -> []
+end
+
+module ElementsById : sig
+  type t
+
+  val empty : t
+
+  val add : [< Identifier.t ] -> [< Component.Element.any ] -> t -> t
+
+  val remove : [< Identifier.t ] -> t -> t
+
+  val find_by_id : Identifier.t -> t -> Component.Element.any option
+end = struct
+  module IdMap = Identifier.Maps.Any
+
+  type t = Component.Element.any IdMap.t
+
+  let empty = IdMap.empty
+
+  let add identifier element t =
+    let check_for_duplicates () =
+      if IdMap.mem (identifier :> Identifier.t) t then
+        Format.eprintf "Duplicate found: %a\n%!" Component.Fmt.model_identifier
+          (identifier :> Identifier.t)
     in
-    (names', IdMap.remove id ids)
+    check_for_duplicates ();
+    IdMap.add (identifier :> Identifier.t) (element :> Component.Element.any) t
 
-  let find_by_name f name (names, _) =
-    let filter e acc = match f e.elem with Some r -> r :: acc | None -> acc in
-    try List.fold_right filter (StringMap.find name names) []
-    with Not_found -> []
+  let remove id t =
+    let id = (id :> Identifier.t) in
+    IdMap.remove id t
 
-  let find_by_id id (_, ids) = try Some (IdMap.find id ids) with _ -> None
+  let find_by_id identifier t =
+    try Some (IdMap.find identifier t) with Not_found -> None
 end
 
 type t = {
   linking : bool;
   (* True if this is a linking environment - if not,
      we only put in modules, module types, types, classes and class types *)
   id : int;
-  elts : Elements.t;
+  elts : ElementsByName.t;
+      (** Elements mapped by their name. Queried with {!find_by_name}. *)
+  ids : ElementsById.t;
+      (** Elements mapped by their identifier. Queried with {!find_by_id}. *)
```

Comment:
This is just a code improvement and not necessary. It was used by an other failed attempt at fixing the issue and I kept the commit.